### PR TITLE
feat(router) Add 'skipLocationChange' and 'replaceUrl' to NavigationEnd event object #22177

### PR DESCRIPTION
**Pull request by passionintellectual**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently, router events such as NavigationEnd does not provide us the details when navigation is done with 'skipLocationChange' or 'replaceUrl' flags
Issue Number: #22177

## What is the new behavior?
As per new behavior, event of type NavigationEnd has one additional parameter called extras with 'skipLocationChange' or 'replaceUrl' flags.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
This is my first pull request to angular repository. Please guide me to the doc page where I should make the changes.

### CODE REVIEW

**Comment by passionintellectual** [comment]
I signed it!

**Comment by matt-senseye** [comment]
I've come across a situation where we need to know whether or not `replaceUrl` was true in the `NavigationEnd` event - it seems like this would be really helpful, does anybody know of another workaround if this PR is not progressing?

File: packages/router/test/integration.spec.ts [hunk 1]
```diff
@@ -477,6 +477,43 @@ describe('Integration', () => {
...
+
+  it('should provide replaceUrl in the NavigationEnd event when changed with navigateByUrl',
+    fakeAsync(inject([Router, Location], (router: Router) => {
+      const fixture = TestBed.createComponent(RootCmp);
+      router.resetConfig([{path: 'team/:id', component: TeamCmp}]);
+      router.events.subscribe((event: RouterEvent)=>{
+        if(event instanceof  NavigationEnd){
+          const extras = (event as NavigationEnd).extras;
+          expect(extras).toBeDefined();
+          if(extras) {
```
Comment (atscott) [inline packages/router/test/integration.spec.ts]: remove this if statement and use the null assertion operator (`!`) if the compiler is complaining about extras possibly being null (same for the previous test).

**Review by atscott — COMMENTED** [review]
Hi @passionintellectual, please reformat this PR with `yarn gulp format` and rebase with the current master.

File: packages/router/test/integration.spec.ts (hunk 1 above)
Comment (passionintellectual) [reply packages/router/test/integration.spec.ts]: Sure, I will do that.

**Comment by atscott** [comment]
FYI - rebased, reformatted and moved `NavigationExtras` to `config.ts` from `router.ts` to avoid cyclic imports between `events.ts` and `router.ts`.

**Comment by atscott** [comment]
[presubmit](http://test/OCL:317156941:BASE:317196901:1592520305019:b7209007)

**Comment by atscott** [comment]
Adding the breaking change label - this can/will break tests that make direct `NavigationEnd` comparisons, i.e.       
```
expect(endEvent).toEqual(new NavigationEnd(1, '/about', '/about'));
```
It might be better to just make the `extras` required so these can be found more easily.

**Comment by atscott** [comment]
> I've come across a situation where we need to know whether or not `replaceUrl` was true in the `NavigationEnd` event - it seems like this would be really helpful, does anybody know of another workaround if this PR is not progressing?

If you subscribe to the events and filter to `NavigationEnd`, I believe you can call [getCurrentNavigation](https://angular.io/api/router/Router#getCurrentNavigation) to get information about the navigation that just completed.

Even though the `currentNavigation` is cleared immediately after the event is emitted, the subscriptions should still process first.
https://github.com/angular/angular/blob/5c40fd65fa89adcd879ae1496683a7f5bdd543e3/packages/router/src/router.ts#L1113-L1117

minimal repro: https://stackblitz.com/edit/rxjs-zdqcnk 
Using the actual router: https://stackblitz.com/edit/angular-ivy-router-base-mcueuc

Because this would be a breaking change, seems somewhat uncommon, and has a simple an alternative available that doesn't require much extra work, I think we might close this PR and feature request.

**Comment by atscott** [comment]
Closing, as this can be accomplished with existing APIs. It's not a priority at this time to invest in a migration to cover this case (especially since the information is already available, as described above). Leaving the bug open for now in case we want to revisit this in the future.

**Pull request closed without merging**